README.md
# eos Cookbook for Arista EOS

[![Build Status](https://jenkins-01.eastus.cloudapp.azure.com/job/arista-eos-cookbook/badge/icon)](https://jenkins-01.eastus.cloudapp.azure.com/job/arista-eos-cookbook/)

The eos cookbook simplifies management of [Arista](https://www.arista.com/) EOS
network devices.  Arista EOS uses the standard el6 32-bit Chef client.  By
including the eos::default recipe in the runlist, it will perform the
following actions needed on EOS:
- Relocate /etc/chef to /persist/sys/chef with a symlink back to /etc/chef.
  This ensures local configuration and state will be retained after a reload.
- Enable eAPI (‘management api http-commands’) with unix-sockets as the
  transport in the running-config
- Add/enhance several ohai plugins

# Requirements

This cookbook is designed and tested with Chef 12 and EOS 4.15, 4.16, and 4.17.
Other versions are likely to work but may not be fully tested.

  - Tested on Arista EOS 4.17 and 4.19
  - Chef client 32-bit RPM for RedHat/CentOS/el
  - Arista Ruby client for eAPI (rbeapi) rubygem and dependencies:
    - rbeapi 1.0 or greater which requires:
      - netaddr
      - net_http_unix
      - inifile

# Installing

Installing Chef on an Arista switch requires the following steps. While the
manual steps are displayed, below, for reference, it is suggested to use a tool
such as Arista CloudVision or ZTP Server to take advantage of the zero-touch
provisioning capability of Arista devices to load a desired EOS version,
additional packages, and a base config, automatically.

- [Download the Chef client](https://downloads.chef.io/chef-client/redhat/)
  for RedHat/CentOS (32-bit)
- Copy the rpm to the switch.

    ```
    Arista#copy http://my_server/path/chef-<version>-1.el6.i386.rpm extension:
    ```

- Install the RPM:

    ```
    Arista#extension chef-<version>-1.el6.i386.rpm
    ```

- Configure EOS to install the chef-client automatically after reloads

    ```
    Arista#copy installed-extensions boot-extensions
    ```

- Ensure `recipe[‘eos’]` is in the default runlist for EOS devices

## Installing behind a firewall

By default, the chef_gem resource will reach out to rubygems.org to find the
necessary rubygems.  When installing on devices without access to the Internet,
additional steps are required.  These, too, should be automated whenever
possible.

Two common solutions are to bootstrap the device with an rbeapi extension or let chef_gem manage the rubygems, directly.

### Install rbeapi as an EOS extension

- [Download the rbeapi extension](https://github.com/arista-eosplus/rbeapi/releases)
  named `rbeapi-chef-<version>.swix`
- Copy the extension to the switch.

    ```
    Arista#copy http://my_server/path/rbeapi-chef-1.0-1.swix extension:
    ```

- Install the extension:

    ```
    Arista#extension rbeapi-chef-1.0-1.swix
    ```

- Configure EOS to install rbeapi automatically after reloads

    ```
    Arista#copy installed-extensions boot-extensions
    ```

### Install rbeapi with chef_gem

Download the rubygem binaries to be included in the cookbook, then use a recipe
to install those on devices.  Example:

Download the rubygem binaries:

```
gem fetch inifile
gem fetch netaddr
gem fetch net_http_unix
gem fetch rbeapi
```

Then, create a recipe to copy these files to nodes and install the packages:

```
cookbook_file “#{Chef::Config[:file_cache_path]}/rbeapi.gem” do
  source ‘rbeapi-1.0.gem’
end
resources(:cookbook_file => “#{Chef::Config[:file_cache_path]}/rbeapi.gem”).run_action(:create)

chef_gem ‘rbeapi’ do
  source “#{Chef::Config[:file_cache_path]}/rbeapi.gem”
  version ‘1.0’
  compile_time false
  action :upgrade
end
```

NOTE: the chef_gem resource requires the `version` to be specified when
installing from a local file.

Finally, include that recipe in the EOS device’s default runlist.
`recipe[eos::rbeapi_local]`

# Using

There are 2 general methods to use this cookbook to manage an Arista switch:
Managing the entire config as a whole or using discrete resources. The
eos_switchconfig resource manages the running-config from a template or file.
Discrete resources, such as eos_vlan, provide selective, granular management of
individual components. Eos_switchconfig is the recommended method for most
network teams. However, eos_vlan is provided to serve as an example for
additional discrete resources to be managed, if desired.

## eos_switchconfig

```ruby
eos_switchconfig 'running-config' do
  action :create
  source 'eos_config.erb'
  variables({
    hostname: 'veos01',
    domainname: 'example.com',
    nameservers: ['10.0.2.3'],
    ntp_server: '10.0.2.3',
    ntp_source_intf: 'Management1',
    static_routes: {
      '0.0.0.0/0' => '10.0.2.2'
    },
    l3ports: [
      Ethernet1: {
        ip_addr: '192.168.8.2/24'
      }
    ],
    l2ports: [
      Ethernet2: {},
      Ethernet3: {},
      Ethernet4: {}
    ]
  })
end
```

## eos_vlan

```ruby
eos_vlan '1' do
  vlan_name 'default'
  enable false
end

eos_vlan '100' do
  vlan_name 'Test_vlan'
  enable true
  trunk_groups %w(mlag_ctl test)
end
```

For more examples, see the [test recipes](test/cookbooks/eos_test/recipes).

# Contributing

Community contributions are welcome.  Please ensure all pull-requests include
spec tests. See [contributing](CONTRIBUTING.md) and [testing](TESTING.md) for more detail.

# Authors & Support

For support, please open a GitHub issue.  This module is maintained by Arista
[EOS+ Consulting Services](mailto://eosplus-dev@arista.com). Commercial support
options are available upon request.

# License

All files in this package are covered by the included BSD 3-clause
[license](LICENSE) unless otherwise noted.


CHANGELOG.md
# Change Log

## [1.0.3](https://github.com/aristanetworks/chef-eos/tree/1.0.3) (2016-11-28)
[Full Changelog](https://github.com/aristanetworks/chef-eos/compare/v1.0.2...1.0.3)

**Merged pull requests:**

- Breakout testing information to TESTING.md [\#9](https://github.com/aristanetworks/chef-eos/pull/9) ([jerearista](https://github.com/jerearista))

## [v1.0.2](https://github.com/aristanetworks/chef-eos/tree/v1.0.2) (2016-11-03)
[Full Changelog](https://github.com/aristanetworks/chef-eos/compare/v1.0.1...v1.0.2)

**Merged pull requests:**

- Doc-updates [\#7](https://github.com/aristanetworks/chef-eos/pull/7) ([jerearista](https://github.com/jerearista))
- Update README.md [\#6](https://github.com/aristanetworks/chef-eos/pull/6) ([jjasghar](https://github.com/jjasghar))

## [v1.0.1](https://github.com/aristanetworks/chef-eos/tree/v1.0.1) (2016-10-10)
[Full Changelog](https://github.com/aristanetworks/chef-eos/compare/v1.0...v1.0.1)

**Merged pull requests:**

- Update metadata and CHANGELOG [\#5](https://github.com/aristanetworks/chef-eos/pull/5) ([jerearista](https://github.com/jerearista))

## [v1.0](https://github.com/aristanetworks/chef-eos/tree/v1.0) (2016-10-10)
[Full Changelog](https://github.com/aristanetworks/chef-eos/compare/0.1.4...v1.0)

**Implemented enhancements:**

- Final updates including Jenkinsfile [\#2](https://github.com/aristanetworks/chef-eos/pull/2) ([jerearista](https://github.com/jerearista))

**Merged pull requests:**

- Release v1.0 [\#4](https://github.com/aristanetworks/chef-eos/pull/4) ([jerearista](https://github.com/jerearista))
- Release 1.0 [\#3](https://github.com/aristanetworks/chef-eos/pull/3) ([jerearista](https://github.com/jerearista))
- Final pre-release cleanup [\#1](https://github.com/aristanetworks/chef-eos/pull/1) ([jerearista](https://github.com/jerearista))

## 0.1.4 (2013-11-26)

- updated providers to support new attributes in Devops Extension

## 0.1.3 (2013-09-04)

- updated provider to use the Devops Extension for EOS

## 0.1.2 (2013-07-19)

- fixed bug found with portchannel data bag entries
- added some additional debug messages to aid with troubleshooting
- change ohai plugin attributes from default to normal for plugin path
- fixed minor issue with get_config to find the right attribute

## 0.1.1 (2013-07-02)

- fixed bug found in edit_portchannel in lag provider
- added initial support for ohai
- added ohai plugin (eos/version)

## 0.1.0 (2013-07-01)

- Initial release of eos cookbook
- Initial release of interface (resource/provider/recipe)
- Initial release of vlan (resource/provider/recipe)
- Initial release of switchport (resource/provider/recipe)
- Initial release of portchannel (resource/provider/recipe)




\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

resources/switchconfig.rb
# Copyright (c) 2016, Arista Networks, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#  * Redistributions of source code must retain the above copyright notice,
#  this list of conditions and the following disclaimer.
#
#  * Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#  and/or other materials provided with the distribution.r
#
#  * Neither the name of Arista Networks nor the names of its contributors may
#  be used to endorse or promote products derived from this software without
#  specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL ARISTA NETWORKS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
# ############################################################################

# eos_switchconfig resource
#
# Note: resource name must always be 'running-config' and
# sub-blocks must be indented with the normal 3-spaces.
#
# Examples:
#   config = '
#   hostname someswitch
#   !
#   interface Ethernet 4
#      description This managed by Chef
#      switchport mode access
#      switchport access vlan 999
#      shutdown
#   !'

#   eos_switchconfig 'running-config' do
#     content config
#     action :create
#   end
#
#   eos_switchconfig 'running-config' do
#     action :create
#     source 'eos_config.erb'
#     variables({
#       hostname: 'veos01',
#       domainname: 'example.com',
#       nameservers: ['10.0.2.3'],
#       ntp_server: '10.0.2.3',
#       ntp_source_intf: 'Management1',
#       static_routes: {
#         '0.0.0.0/0' => '10.0.2.2'
#       },
#       l3ports: [
#         Ethernet1: {
#           ip_addr: '172.16.130.181/24'
#         }
#       ],
#       l2ports: [
#         Ethernet2: {},
#         Ethernet3: {},
#         Ethernet4: {}
#       ]
#     })
#   end
#

property :config_file, String, name_property: true
property :switch_name, String, desired_state: false
property :content, String, required: false
property :file_name, String, required: false
property :source, String, required: false, desired_state: true
property :variables, Hash, required: false, desired_state: true
property :force, kind_of: [TrueClass, FalseClass], default: false,
                 desired_state: false

default_action :create

require 'pry'
begin
  require 'rbeapi'
  require 'rbeapi/switchconfig'
rescue LoadError
  msg = 'Unable to load rbeapi rubygem'
  Chef::Log.debug msg
end

##
# Instance of Rbeapi::Client::switch used to sending and receiving
# eAPI messages.  In addition, the switch object provides access to
# Ruby Client for eAPI API modules used to configure EOS resources.
#
# @return [switch] An instance of Rbeapi::Client::switch used to send
#   and receive eAPI messages.
def switch
  return @switch if @switch
  Rbeapi::Client.load_config(ENV['RBEAPI_CONF']) if ENV['RBEAPI_CONF']
  connection_name = ENV['RBEAPI_CONNECTION'] || switch_name || 'localhost'
  @switch = Rbeapi::Client.connect_to(connection_name)
end

##
# Perform a block-by-clock diff of the current and desired configs.
# This allows for user input in which blocks are not in the same order
# as 'show running-config'.
#
# @param [current] <String> The running-config from the switch.
# @param [desired] <String> The proposed config from the user.
# @return [Boolean] whether configurations differ
# rubocop:disable AbcSize
def configs_differ?(current, desired)
  # Get the current running config in a SwitchConfig object
  org_swc = Rbeapi::SwitchConfig::SwitchConfig.new(current)

  # Get the new running config in a SwitchConfig object
  new_swc = Rbeapi::SwitchConfig::SwitchConfig.new(desired)

  # Compare the current and new configs
  # If results are both empty then nothing needs to change,
  results = org_swc.compare(new_swc)
  !results[0].cmds.empty? || \
    !results[0].children.empty? || \
    !results[1].cmds.empty? || \
    !results[1].children.empty?
end
# rubocop:enable AbcSize

load_current_value do |desired_resource|
  # Get the current values from the switch
  current = switch.get_config(config: 'running-config', as_string: true)

  if desired_resource.source
    source desired_resource.source
    # A template was passed to us
    require 'chef/mixin/template'
    @template_context = Chef::Mixin::Template::TemplateContext.new({})

    # Get the file path to the template in the cookbook
    templates = run_context.cookbook_collection[desired_resource.cookbook_name]
                           .template_filenames
    source_path = ''
    templates.each do |tpath|
      source_path = tpath if tpath =~ /#{desired_resource.source}(.erb)?$/
    end
    if source_path.empty?
      Chef::Log.fatal "Unable to locate template: #{desired_resource.source}"
    end

    # Copy any variables in to the template context
    if desired_resource.variables
      variables desired_resource.variables
      desired_resource.variables.each do |key, value|
        @template_context[key] = value
      end
    end

    # Render the template to a string
    desired_resource.content @template_context.render_template(source_path)
  elsif desired_resource.file_name
    file_name desired_resource.file_name

    # Get the file path in the cookbook
    files = run_context.cookbook_collection[desired_resource.cookbook_name]
                       .file_filenames
    source_path = ''
    files.each do |fpath|
      source_path = fpath if fpath.end_with? desired_resource.file_name
    end
    if source_path.empty?
      Chef::Log.fatal "Unable to locate file: #{desired_resource.source}"
    end
    desired_resource.content IO.read(source_path)
  end

  # Use the block-compare to determine if the configs are different
  if !configs_differ?(current, desired_resource.content)
    content desired_resource.content
  else
    content current
  end
end

action :create do
  converge_if_changed do
    eos_v = switch.enable(['show version'])[0][:result]['version']
    # Introduced in 4.14.6M, Recommended for use in 4.15.0F
    if Gem::Version.new(eos_v) > Gem::Version.new('4.15')
      if new_resource.content
        file '/mnt/flash/startup-config' do
          content new_resource.content
        end
      elsif new_resource.resource.eql?(:cookbook_file)
        cookbook_file '/mnt/flash/startup-config' do
          cookbook new_resource.cookbook
          source new_resource.source_file
        end
      elsif new_resource.resource.eql?(:template)
        template '/mnt/flash/startup-config' do
          cookbook new_resource.cookbook
          source new_resource.source_file
          variables new_resource.variables
        end
      end

      execute 'replace running-config' do
        command 'FastCli -p15 -c "configure replace flash:startup-config"'
      end

    else
      Chef::Log.fatal 'Config replace requires EOS version 4.15 or higher.'
    end
  end
end
